src/toolchain.ml

```diff
@@ -181,14 +181,73 @@ let of_settings () : t option Promise.t =
           (Opam.Switch.name switch);
         None
       ) )
-  | Some Global -> Promise.return (Some Global)
+  | Some System -> Promise.return (Some System)
   | Some (Custom template) -> Promise.return (Some (Custom template))
 
+let detect_esy_sandbox ~project_root esy () =
+  let open Promise.Syntax in
+  let* esy = esy in
+  let* esy_build_dir_exists = Fs.exists (Filename.concat project_root "_esy") in
+  let* manifest = Esy.find_manifest_in_dir project_root in
+  match (esy, esy_build_dir_exists, manifest) with
+  | Some esy, true, Some manifest ->
+    (* Some projects have both use both Esy and Opam, we can't assume that the user
+       wants to use Esy just by looking for an "esy.json" file.
+
+       Instead, we assume the user wants to use esy if these three conditions are met:
+        - The esy command is in the user's system
+        - There is an _esy build directory in the project root
```

Comment:
Condition 2 is actually the only one that would allow us to guess that the user probably wants to use Esy:

- Esy can be used with `package.json`, `esy.json`, or even without any of those and just use the local `.opam` files.
- Lots of projects (e.g. `vscode-ocaml-platform`) support using both Opam and Esy (so we can't assume the user uses Esy simply by checking if the project has an `esy.json` file).

Since it seems difficult to know what the user wants to use by just looking at the files in the project, I think a reliable way of guessing it is to check if the user compiled the project with Esy (which will always generate an `_esy` folder).

A drawback of this approach is that the Esy sandbox won't be selected if the user just open the project for the first time, before compiling it. But in this scenario, it does not seem possible to predict that the user wants to use Esy, and I think we want to prioritize Opam.

Reply:
Yeah, that's how I viewed things initially as well, but now I'm thinking that it's a mistake to be this conservative.

As you've pointed, the main utility of this feature is when users initially clone a project to get started. This is precisely when condition 2 will not hold. After they've built their projects, it's likely that they've saved their choice in in settings file and no longer care for this detection.

Second of all, we need to consider two pools of users that have installed both esy & opam:
* The group with a strong preference for opam
* The group with a strong preference for esy

I would guess that the 2nd group is much longer, but our extension is mostly optimizing for the 1st group.

To me what seems reasonable is to auto detect and use esy whenever we know it's quite likely the first choice (existence of esy binary + esy.json). And if we hear complaints, we can always add a package manager preference option.

In any case, I don't use esy don't have the motivation to improve things for esy users. So I will not let this block the PR.

Reply:
I agree that the scenario to optimize for is when the user opens the project for the first time, and with the proposed approach, we completely fail Esy users for this.

I don't use Esy, so I don't have a strong opinion on when we should choose to use it, as long as Esy users are happy.

However, in order to keep the first group (Opam users) happy, I don't think simply looking for an `esy.json` file is conservative enough: I don't use Esy, but have it in my system because I use it for Spin's integration tests. So with the approach you propose, the sandbox detection wouldn't detect my preference correctly for `vscode-ocaml-platform`, for instance. And I don't think my case is isolated: I figure lots of Opam users have installed Esy, used it once or twice, but will prefer their detected package manager to be Opam.

So it's really a question of where we put the threshold: either we introduce friction for an (unknown) part of Opam users, or we introduce friction for the entirety of Esy userbase. I would lean for the latter, but I'm obviously biased. Will let you make the call: that was my last argument 😄

Reply:
Let's leave it as is. If an esy user will care enough they will send a PR.